src/ledger/backend/ethereumOnChainQuerier.spec.ts

```diff
@@ -0,0 +1,165 @@
+	let alice: Entity;
+	let bob: Entity;
+	let onChainQueryA: OnChainQuerier;
+	let aTransferSeed: number;
+	let bTransferSeed: number;
+
+	const initEntity = async (wallet: Wallet): Promise<Entity> => {
+		const tokens = test.newRandomTokens(RNG, TOKEN_SIZE);
+		const entity: Entity = {
+			wallet,
+			tokens,
+			runningTokens: Array.from(tokens.value),
+		};
+
+		const erc721Minter = IERC721Minter__factory.connect(
+			testenv.perunArt,
+			entity.wallet,
+		);
+		let nonce = await entity.wallet.provider.getTransactionCount(
+			entity.wallet.address,
+		);
+		for (const id of entity.tokens.value) {
+			await erc721Minter.mint(entity.wallet.address, id, {
+				nonce: nonce++,
+			});
+		}
+		await testenv.mine();
+		// await Promise.all(
+		// 	entity.tokens.value.map((id) => {
+		// 		erc721Minter.mint(entity.wallet.address, id, {
+		// 			nonce: nonce++,
+		// 		});
+		// 	}),
+		// );
+
+		return entity;
+	};
+
+	const seededTokenTransfer = async (
+		from: Entity,
+		to: Entity,
+		seed: number,
+		tokenPool: Tokens,
+		forwardChain: boolean,
+	): Promise<void> => {
+		const ierc721 = IERC721__factory.connect(testenv.perunArt, from.wallet);
+		let nonce = await from.wallet.provider.getTransactionCount(
+			from.wallet.address,
+		);
+		for (let index = 0; index < TOKEN_SIZE; ++index) {
+			if ((seed % PRIMES[index]) % 2 === 0) {
+				const token = tokenPool.value[index];
+				await ierc721.transferFrom(
+					from.wallet.address,
+					to.wallet.address,
+					token,
+					{ nonce: nonce++ },
+				);
+				to.runningTokens.push(token);
+				from.runningTokens.splice(from.runningTokens.indexOf(token), 1);
+				if (forwardChain) {
+					await testenv.mine();
+				}
+			}
+		}
+		await testenv.mine();
+	};
+
+	before(async function () {
+		testenv = await setupEnv(2, undefined, { blockTime: 1 });
+		alice = await initEntity(testenv.users[0]);
+		bob = await initEntity(testenv.users[1]);
+		onChainQueryA = new EthereumOnChainQuerier(alice.wallet);
+	});
+
+	it("correctly returns the tokens in the address' possession after transfers across blocks", async function () {
+		aTransferSeed = RNG.uInt32();
+		await seededTokenTransfer(
+			alice,
+			bob,
+			aTransferSeed,
+			alice.tokens,
+			true,
+		);
+		bTransferSeed = RNG.uInt32();
+		await seededTokenTransfer(bob, alice, bTransferSeed, bob.tokens, true);
+
+		let aliceTokens = await onChainQueryA.queryTokensOwnedByAddress(
+			testenv.perunArt,
+			alice.wallet.address,
+		);
+		expect(aliceTokens.value).to.have.members(alice.runningTokens);
+		expect(aliceTokens.value).to.have.length(alice.runningTokens.length);
+		let bobTokens = await onChainQueryA.queryTokensOwnedByAddress(
+			testenv.perunArt,
+			bob.wallet.address,
+		);
+		expect(bobTokens.value).to.have.members(bob.runningTokens);
+		expect(bobTokens.value).to.have.length(bob.runningTokens.length);
+	});
+
+	it("correctly returns the tokens in the address' possession after single-block transfers", async function () {
```

Comment:
If I see correctly, this test and the prior are the exact same, besides the boolean `forwardChain` being set to true and false. In such cases, please always move the code into a function and then call it with the different parameters (here, `true` and `false`). D.R.Y.